File: src/backend/utils/rate_limit.py
import time
from collections import defaultdict
from dataclasses import dataclass
from typing import Dict, List
import os
import json
from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware
from .logger import logger
from src.backend.llm.keycheck import CREDITS_AVAILABLE

@dataclass
class RateLimit:
    max_requests: int
    window_seconds: int
    requests: List[float] = None

    def __post_init__(self):
        self.requests = []

    def is_allowed(self) -> bool:
        """Check if request is allowed within rate limit"""
        now = time.time()
        # Remove old requests
        self.requests = [t for t in self.requests if now - t < self.window_seconds]
        return len(self.requests) < self.max_requests

    def add_request(self):
        """Add a request to the list"""
        now = time.time()
        # Clean expired requests first
        self.requests = [t for t in self.requests if now - t < self.window_seconds]
        self.requests.append(now)

    def get_remaining(self) -> int:
        """Get remaining requests"""
        now = time.time()
        # Only count requests that aren't expired
        valid_requests = [t for t in self.requests if now - t < self.window_seconds]
        return self.max_requests - len(valid_requests)

class RateLimiter:
    def __init__(self):
        # Get limits from environment or use defaults
        self.hourly_limit = int(os.getenv("RATE_LIMIT_PER_HOUR", "15"))
        self.daily_limit = int(os.getenv("RATE_LIMIT_PER_DAY", "50"))
        
        # Initialize counters
        self.hourly_requests = defaultdict(list)  # {ip: [timestamps]}
        self.daily_requests = defaultdict(list)   # {ip: [timestamps]}
        self.request_ids = set()  # Track unique request IDs
        
        # Current time window
        self.hour = 3600  # seconds
        self.day = 86400  # seconds
        
    def _cleanup_old_requests(self, ip: str):
        """Remove expired timestamps"""
        current_time = time.time()
        
        # Cleanup hourly
        self.hourly_requests[ip] = [
            ts for ts in self.hourly_requests[ip]
            if current_time - ts < self.hour
        ]
        
        # Cleanup daily
        self.daily_requests[ip] = [
            ts for ts in self.daily_requests[ip]
            if current_time - ts < self.day
        ]
        
        # Cleanup old request IDs (older than a day)
        self.request_ids = {
            rid for rid in self.request_ids
            if rid.split('-')[0].startswith('t')  # Only keep our timestamped IDs
            and current_time - float(rid.split('-')[0][1:]) < self.day  # Remove 't' prefix
        }

    def is_allowed(self, ip: str, request_id: str = None) -> bool:
        """Check if request is allowed based on rate limits"""
        # Skip rate limiting for Ollama requests (when credits not available)
        if not CREDITS_AVAILABLE:
            return True
            
        self._cleanup_old_requests(ip)
        
        # If request_id provided, check if it's a duplicate
        if request_id:
            if request_id in self.request_ids:
                return True  # Allow duplicate requests (retries)
            # Store with timestamp prefix for cleanup
            timestamped_id = f"t{time.time()}-{request_id}"
            self.request_ids.add(timestamped_id)
        
        # Check limits
        if len(self.hourly_requests[ip]) >= self.hourly_limit:
            return False
            
        if len(self.daily_requests[ip]) >= self.daily_limit:
            return False
        
        # Add timestamp for new request
        current_time = time.time()
        self.hourly_requests[ip].append(current_time)
        self.daily_requests[ip].append(current_time)
        
        return True
    
    def get_remaining(self, ip: str) -> dict:
        """Get remaining requests for the client"""
        self._cleanup_old_requests(ip)
        
        # If using Ollama, return unlimited
        if not CREDITS_AVAILABLE:
            return {
                "hourly_remaining": 999,
                "daily_remaining": 999
            }
        
        return {
            "hourly_remaining": max(0, self.hourly_limit - len(self.hourly_requests[ip])),
            "daily_remaining": max(0, self.daily_limit - len(self.daily_requests[ip]))
        }

# Create singleton instance
rate_limiter = RateLimiter()

class RateLimitMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next) -> Response:
        # Get client IP
        client_ip = request.client.host
        
        # Only check and enforce rate limits for LLM API calls
        is_llm_api = request.url.path.startswith("/llm") and request.method == "POST"
        
        # Get request ID from header if present (for retries/backup attempts)
        request_id = request.headers.get("X-Request-ID")
        
        if is_llm_api and CREDITS_AVAILABLE:  # Only rate limit when using OpenRouter
            if not rate_limiter.is_allowed(client_ip, request_id):
                remaining = rate_limiter.get_remaining(client_ip)
                error_response = {
                    "error": {
                        "code": 429,
                        "message": "Rate limit exceeded",
                        "details": {
                            "hourly_remaining": remaining["hourly_remaining"],
                            "daily_remaining": remaining["daily_remaining"],
                            "hourly_limit": rate_limiter.hourly_limit,
                            "daily_limit": rate_limiter.daily_limit,
                            "retry_after": "Wait for the next hour or day depending on which limit was exceeded"
                        }
                    }
                }
                return Response(
                    content=json.dumps(error_response),
                    status_code=429,
                    headers={
                        "Content-Type": "application/json",
                        "X-RateLimit-Remaining-Hour": str(remaining["hourly_remaining"]),
                        "X-RateLimit-Remaining-Day": str(remaining["daily_remaining"])
                    }
                )
        
        # Process the request
        response = await call_next(request)
        
        # Add rate limit headers to response
        if is_llm_api:
            remaining = rate_limiter.get_remaining(client_ip)
            response.headers["X-RateLimit-Remaining-Hour"] = str(remaining["hourly_remaining"])
            response.headers["X-RateLimit-Remaining-Day"] = str(remaining["daily_remaining"])
        
        return response 

File: src/backend/utils/rate_limit.ts
import { FastifyRequest, FastifyReply } from 'fastify';
import { config } from '../config';
import { logger } from './logger';

interface RequestWindow {
  timestamps: number[];
}

interface RateLimitResponse {
  hourlyRemaining: number;
  dailyRemaining: number;
}

class RateLimiter {
  private hourlyRequests: Map<string, number[]>;
  private dailyRequests: Map<string, number[]>;
  private requestIds: Set<string>;
  private readonly HOUR = 3600; // seconds
  private readonly DAY = 86400; // seconds
  private creditsCheck: () => boolean;

  constructor() {
    this.hourlyRequests = new Map();
    this.dailyRequests = new Map();
    this.requestIds = new Set();
    this.creditsCheck = () => true;
  }

  private cleanupOldRequests(ip: string): void {
    const currentTime = Date.now() / 1000;

    // Cleanup hourly requests
    const hourlyTimestamps = this.hourlyRequests.get(ip) || [];
    this.hourlyRequests.set(
      ip,
      hourlyTimestamps.filter(ts => currentTime - ts < this.HOUR)
    );

    // Cleanup daily requests
    const dailyTimestamps = this.dailyRequests.get(ip) || [];
    this.dailyRequests.set(
      ip,
      dailyTimestamps.filter(ts => currentTime - ts < this.DAY)
    );

    // Cleanup old request IDs
    this.requestIds = new Set(
      Array.from(this.requestIds).filter(rid => {
        const [timestamp] = rid.split('-');
        return timestamp.startsWith('t') &&
          currentTime - parseFloat(timestamp.slice(1)) < this.DAY;
      })
    );
  }

  setCreditsCheck(callback: () => boolean): void {
    this.creditsCheck = callback;
  }

  isAllowed(ip: string, requestId?: string): boolean {
    // Skip rate limiting for backup LLM
    if (!this.creditsCheck()) {
      return true;
    }

    this.cleanupOldRequests(ip);

    // Handle duplicate requests
    if (requestId) {
      if (this.requestIds.has(requestId)) {
        return true;
      }
      const timestampedId = `t${Date.now() / 1000}-${requestId}`;
      this.requestIds.add(timestampedId);
    }

    const hourlyCount = (this.hourlyRequests.get(ip) || []).length;
    const dailyCount = (this.dailyRequests.get(ip) || []).length;

    return hourlyCount < config.rateLimiting.hourlyLimit &&
           dailyCount < config.rateLimiting.dailyLimit;
  }

  addRequest(ip: string): void {
    const currentTime = Date.now() / 1000;
    
    // Initialize arrays if they don't exist
    if (!this.hourlyRequests.has(ip)) {
      this.hourlyRequests.set(ip, []);
    }
    if (!this.dailyRequests.has(ip)) {
      this.dailyRequests.set(ip, []);
    }

    // Add timestamps
    this.hourlyRequests.get(ip)?.push(currentTime);
    this.dailyRequests.get(ip)?.push(currentTime);
  }

  getRemaining(ip: string): RateLimitResponse {
    this.cleanupOldRequests(ip);

    // Return unlimited for backup LLM
    if (!this.creditsCheck()) {
      return {
        hourlyRemaining: 999,
        dailyRemaining: 999
      };
    }

    const hourlyCount = (this.hourlyRequests.get(ip) || []).length;
    const dailyCount = (this.dailyRequests.get(ip) || []).length;

    return {
      hourlyRemaining: Math.max(0, config.rateLimiting.hourlyLimit - hourlyCount),
      dailyRemaining: Math.max(0, config.rateLimiting.dailyLimit - dailyCount)
    };
  }
}

// Create singleton instance
export const rateLimiter = new RateLimiter();

// Fastify plugin for rate limiting
export async function rateLimitPlugin(
  request: FastifyRequest,
  reply: FastifyReply
) {
  const clientIp = request.ip;
  const requestId = request.headers['x-request-id'] as string;

  // Only check LLM generation endpoints
  const isLLMGeneration = (
    request.url.startsWith('/llm/pace-notes/generate') ||
    request.url.startsWith('/llm/policyfoo/generate')
  ) && request.method === 'POST';

  if (isLLMGeneration && rateLimiter.creditsCheck()) {
    if (!rateLimiter.isAllowed(clientIp, requestId)) {
      const remaining = rateLimiter.getRemaining(clientIp);
      return reply.status(429).send({
        error: {
          code: 429,
          message: 'Rate limit exceeded',
          details: {
            hourlyRemaining: remaining.hourlyRemaining,
            dailyRemaining: remaining.dailyRemaining,
            hourlyLimit: config.rateLimiting.hourlyLimit,
            dailyLimit: config.rateLimiting.dailyLimit,
            retryAfter: 'Wait for the next hour or day depending on which limit was exceeded'
          }
        }
      });
    }
  }

  // Add response hook to track successful requests
  reply.addHook('onSend', (request, reply, payload, done) => {
    if (isLLMGeneration && reply.statusCode === 200) {
      rateLimiter.addRequest(clientIp);
    }
    
    // Add rate limit headers
    const remaining = rateLimiter.getRemaining(clientIp);
    reply.header('X-RateLimit-Remaining-Hour', remaining.hourlyRemaining.toString());
    reply.header('X-RateLimit-Remaining-Day', remaining.dailyRemaining.toString());
    reply.header('Access-Control-Expose-Headers', 'X-RateLimit-Remaining-Hour, X-RateLimit-Remaining-Day');
    
    done(null, payload);
  });
} 